### ISSUE TITLE
Change bytecode compilers to cache at task/operator level

### ISSUE THREAD

**Issue by cberner** [prestodb/presto#6151]
Currently, compilers like JoinCompiler cache 1000 compilations, globally. It would be better to cache compilations at the task/operator level, and then discard them when the query finishes. As it is, old compiled classes can bloat the JVM code cache

**Comment by Praveen2112** [comment]
But when we are caching compilations globally we are getting a better performance rather then compiling each time when a same query is executed.(#6153)

**Comment by cberner** [comment]
Interesting. How did you determine that the speed up was from caching of the generated code?

**Comment by Praveen2112** [comment]
We designed a custom connector which actually fetches data from memory. Presto is running one query at a time so there isnt much a load. And we ran the same query for the second time , time taken is reduced to max of 3 seconds for complex query. When we profiled the max time is taken by each split when it is processing the first page(The time varies from 17ms for first page to 4ms for 5th page having same position count). So our guess is that JVM JIT might optimize it on the run time and optimized code is maintained in code cache and when the same query is executed next time , JIT doesnt have to optimize it again.

Correct me if I am wrong.
